blueprints/cloud-operations/uptime-monitoring-tool/source/net-mon-agent/main.py

```diff
@@ -0,0 +1,160 @@
+    interval = monitoring_v3.TimeInterval(
+          {"end_time": {"seconds": seconds, "nanos": nanos}}
+    )
+
+    uptime_check_series = monitoring_v3.TimeSeries()
+    uptime_check_series.metric.type = f"custom.googleapis.com/{ENDPOINT_UPTIME_METRIC}"
+    uptime_check_series.resource.type = "global"
+    uptime_check_series.metric.labels[
+        "endpoint"] = f"{uptime_check_result.host}:{uptime_check_result.port}"
+    uptime_check_series.metric.labels["agent_vm"] = agent_vm
+    uptime_point = monitoring_v3.Point(
+          {"interval": interval, "value": {"bool_value": uptime_check_result.reachable}})
+    uptime_check_series.points = [uptime_point]
+    time_series.append(uptime_check_series)
+
+    if uptime_check_result.reachable:
+        latency_series = monitoring_v3.TimeSeries()
+        latency_series.metric.type = f"custom.googleapis.com/{ENDPOINT_LATENCY}"
+        latency_series.resource.type = "global"
+        latency_series.metric.labels[
+            "endpoint"] = f"{uptime_check_result.host}:{uptime_check_result.port}"
+        latency_series.metric.labels["agent_vm"] = agent_vm
+        latency_point = monitoring_v3.Point(
+              {"interval": interval, "value": {"int64_value": uptime_check_result.latency}})
+        latency_series.points = [latency_point]
+        time_series.append(latency_series)
+
+    try:
+        client.create_time_series(name=f"projects/{project}", time_series=time_series)
+    except Exception as e:
+        LOGGER.error(f"Failed to write time series data: {e}")
+
+
+@click.command()
+@click.option('--project', '-p', type=str, required=True, help='Target GCP project id for metrics.')
+@click.option('--endpoint', type=(str, int), multiple=True,
+              help='Endpoints to test as ip port whitespace separated, this is a repeatable arg.')
+@click.option('--agent', type=str, required=True, help='Agent VM running the script')
+@click.option('--debug', is_flag=True, default=False,
+              help='Turn on debug logging.')
+@click.option('--interval', type=int, default=10, help='Time interval between checks in seconds.')
+@click.option('--timeout', type=int, default=10,
+              help='Timeout waiting for connection to be established.')
+def main(project: str, endpoint: list, agent: str, debug: bool, interval: int, timeout: int):
+    logging.basicConfig(level=logging.INFO if not debug else logging.DEBUG)
+
+    while True:
+        # Create a list of futures to represent the concurrent tasks.
+        with concurrent.futures.ThreadPoolExecutor() as executor:
+            futures = {executor.submit(check_host_connectivity, timeout, host, port) for
+                       (host, port) in endpoint}
+
+        # Wait for all the tasks to finish.
+        try:
+            for future in concurrent.futures.as_completed(futures):
+                uptime_check_result = future.result()
+                write_uptime_check_results_to_monitoring(project=project, agent_vm=agent,
```

Comment:
you might want to batch up monitoring requests in a single call, a simple implementation is in the quota monitor

Reply:
Aligned batch api requests for publishing time series to quota monitor script